### PR TITLE
Add topic model, migration and tests

### PR DESCRIPTION
**Pull request by AgostinaDev**
Adds part of the Topics CRUD
- Topic model, migration and some tests for this model were added. Image processing has not been added yet.

Topics are part of the Targets and will be managaged by administrators. 
- Trello card: https://trello.com/c/jCC7w7D9

### CODE REVIEW

File: db/migrate/20210324174033_create_topics.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CreateTopics < ActiveRecord::Migration[6.1]
+  def up
+    create_table :topics do |t|
+      t.string "description", :limit => 50, :null => false
+      t.string "image"
```
Comment (federicoaldunate) [inline db/migrate/20210324174033_create_topics.rb]: If you want to have images, maybe you can use `activestorage`.
If you add `activestorage`, there's no need to add it now

File: spec/models/topic_spec.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#  description :string(50)       not null
+#  image       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Topic, type: :model do
+	let!(:topics) do
+		[Topic.create(:description => "Music", :image => "image string 1"), Topic.create(:description => "Sports", :image => "image string 2"), Topic.create(:description => "Food", :image => "image string 3")]
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: Here you can use `create_list`

File: spec/models/topic_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	let(:topic) do
+		Topic.create(:description => "Plants", :image => "image string here")
+	end
+
+	# TODO: See if repetition of 'find_by to/not_to be_nil' expectations should be avoided
+
+  it { should validate_presence_of(:description) }
+
+  it "is expected to save topics successfully after #create" do
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: be careful with indentation, perhaps you can add a tool or gem to help you with this

File: spec/models/topic_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		expect(Topic.all).to match_array(topics)
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: Remember that we want to test the model. Here you might be testing the testing suite.

File: spec/models/topic_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		expect(Topic.count).to eq(3)
+	end
+
+	it "is expected to save description successfully after #update" do
+		topic.description = "Zombies"
+		topic.save
+		expect(Topic.find_by(:description => "Plants")).to be_nil
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: try using the new syntax, that is `Topic.find_by(description: 'Plants')`
For example here you are testing if `saves` works correctly. That is something that is already tested on ActiveRecord class. You inherited from ActiveRecord when creating your model class and you can assume `save` mehtod will work fine.

File: spec/models/topic_spec.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		topic.image = "new image"
+		topic.save
+		expect(Topic.find_by(:image => "new image")).not_to be_nil
+		expect(Topic.find_by(:image => "image string here")).to be_nil
+	end
+
+	it "is expected to delete topic successfully after #destroy" do
+		expect(topic).not_to be_nil
+		topic.destroy
+		expect(Topic.find_by(:id => topic.id)).to be_nil
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: Same here with destroy about testing something out of the scope of your class

File: spec/models/topic_spec.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+#
+#  id          :bigint           not null, primary key
+#  description :string(50)       not null
+#  image       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Topic, type: :model do
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb:12]: What might be a good thing to test (and easy) are your validations. You can test that the validation on `describe` works fine

File: db/migrate/20210324174033_create_topics.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+      t.timestamps
+    end
+  end
+
+  def down
+    drop_table :topics
+  end
+end
```
Comment (TimoPeraza) [inline db/migrate/20210324174033_create_topics.rb]: end of line missing

File: db/migrate/20210324174033_create_topics.rb (hunk 1 above)
Comment (TimoPeraza) [inline db/migrate/20210324174033_create_topics.rb]: here you can write this code in the next way:
```
      t.string :description, limit: 50, null: false
      t.string  :image
```

File: spec/models/topic_spec.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		[Topic.create(:description => "Music", :image => "image string 1"), Topic.create(:description => "Sports", :image => "image string 2"), Topic.create(:description => "Food", :image => "image string 3")]
+	end
+
+	let(:topic) do
+		Topic.create(:description => "Plants", :image => "image string here")
+	end
+
+	# TODO: See if repetition of 'find_by to/not_to be_nil' expectations should be avoided
+
+  it { should validate_presence_of(:description) }
```
Comment (TimoPeraza) [inline spec/models/topic_spec.rb]: Maybe you can put this validation into a context with description "validations"

File: spec/factories/topics.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+#
+#  id          :bigint           not null, primary key
+#  description :string(50)       not null
+#  image       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+FactoryBot.define do
+  factory :topic do
+    trait :music do
```
Comment (TimoPeraza) [inline spec/factories/topics.rb]: maybe, is a good idea add a gem named faker https://github.com/faker-ruby/faker

This gem will help you to add random data in your factories

So, you will be able to add a line of code like this:
``` description { Faker::Lorem.sentence }```

File: app/models/topic.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+# Table name: topics
+#
+#  id          :bigint           not null, primary key
+#  description :string(50)       not null
+#  image       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+class Topic < ApplicationRecord
+  validates :description, presence: true, uniqueness: true
```
Comment (TimoPeraza) [inline app/models/topic.rb]: maybe, is a good idea to add a validation to check the length of the description

File: spec/models/topic_spec.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+require 'rails_helper'
+
+RSpec.describe Topic, type: :model do
+	context 'Validations' do
+		let!(:topics) do
+			create :topic, :music, :videogames, :plants, :sports
+		end
+
+	  it { is_expected.to validate_presence_of(:description) }
+	  it { is_expected.to validate_uniqueness_of(:description) }
```
Comment (TimoPeraza) [inline spec/models/topic_spec.rb]: also, you can add a test to the validation of the length

File: db/migrate/20210324174033_create_topics.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    create_table :topics do |t|
+      t.string 'description', limit: 50, null: false
+
+      t.timestamps
+    end
+  end
+
+  def down
+    drop_table :topics
+  end
```
Comment (TimoPeraza) [inline db/migrate/20210324174033_create_topics.rb]: you don't need to do an up and a down method, just using change method is enough 😄

File: Gemfile [hunk 14]
```diff
@@ -37,8 +37,16 @@ end
 group :development do
   gem 'listen', '~> 3.3'
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
+  gem 'annotate'
```
Comment (TimoPeraza) [inline Gemfile]: is a good practice to indicate the gem version 😄 like the gem above

File: spec/rails_helper.rb [hunk 15]
```diff
@@ -61,4 +62,4 @@
   config.filter_rails_from_backtrace!
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
-end
+end
```
Comment (TimoPeraza) [inline spec/rails_helper.rb]: don't forget end of line 😄

File: spec/support/factory_bot.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    DatabaseCleaner.strategy = :transaction
+  end
+
+  # start the transaction strategy as examples are run
+  config.around(:each) do |example|
+    DatabaseCleaner.cleaning do
+      example.run
+    end
+  end
+end
```
Comment (TimoPeraza) [inline spec/support/factory_bot.rb]: the same here

File: spec/support/shoulda_matchers.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+# configure shoulda matchers to use rspec as the test framework and full matcher libraries for rails
+Shoulda::Matchers.configure do |config|
+  config.integrate do |with|
+    with.test_framework :rspec
+    with.library :rails
+  end
+ end
```
Comment (TimoPeraza) [inline spec/support/shoulda_matchers.rb]: the same here

File: spec/models/topic_spec.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+#  description :string(50)       not null
+#  image       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Topic, type: :model do
+	context 'Validations' do
+	  let!(:topics) { create_list(:topic, 15) }
```
Comment (federicoaldunate) [inline spec/models/topic_spec.rb]: You don't need to create 15 topics.
And when testing validations there is no need to create a topic, you can build one. Creation needs a DB and is more expensive. With few tests, it doesn't matter, but when the project grows we want tests to be as fast as possible

**Review by federicoaldunate — APPROVED** [review]
I'm approving considering you will fix all comments. Nice work! 💪

**Comment by federicoaldunate** [comment]
> I'm approving considering you will fix all comments. Nice work! 💪

I think you need to add me to the team or to the repo `Only comments by reviewers with write access count toward mergeability`

**Pull request merged**